### PR TITLE
KAFKA-2666: Docs: Automatically generate documentation from config classes

### PR DESCRIPTION
**Pull request by gwenshap**
…the way we always planned to

### CODE REVIEW

**Comment by junrao** [comment]
Thanks a lot for the patch. This is super useful! A couple of comments.
1. If the doc tgz file can include javadoc/, that will be great. If it's not easy to do, the patch is fine as it is.
2. Also, ConfigDef.toHtmlTable() seems to have a bug. The doc says first sort by required (which is correct), but it's actually sorting by hasDefault. Could you fix it in this patch too?

File: docs/documentation.html [hunk 1]
```diff
@@ -41,8 +41,9 @@
         <ul>
              <li><a href="#brokerconfigs">3.1 Broker Configs</a>
              <li><a href="#consumerconfigs">3.2 Consumer Configs</a>
-             <li><a href="#producerconfigs">3.3 Producer Configs</a>
-			 <li><a href="#newproducerconfigs">3.4 New Producer Configs</a>
+             <li><a href="#newconsumerconfigs">3.3 New Consumer Configs</a>
+             <li><a href="#producerconfigs">3.4 Kafka Producer Configs</a>
+			 <li><a href="#newproducerconfigs">3.5 Old Producer Configs</a>
```
Comment (junrao) [inline docs/documentation.html]: indentation

**Comment by guozhangwang** [comment]
LGTM overall besides Jun's comments. One general comment: if we are going to make the new producer as the default producer, then we can probably make the new producer docs in the Producer API and Producer Configs section, and list the old producer API / configs as a note as we did for the new producer APIs now instead of a separate subsection.

File: docs/configuration.html [hunk 2]
```diff
@@ -736,8 +316,18 @@
 
 
 <p>More details about consumer configuration can be found in the scala class <code>kafka.consumer.ConsumerConfig</code>.</p>
-<h3><a id="producerconfigs">3.3 Producer Configs</a></h3>
-Essential configuration properties for the producer include:
+
+    <h3><a id="newconsumerconfigs">3.3 New Consumer Configs</a></h3>
+    We are working on a replacement for our existing simple and high-level consumers. The code is available in trunk now and can be considered beta quality. Below is the configuration for the new consumer: </p>
```
Comment (guozhangwang) [inline docs/configuration.html]: We can probably mention "Since 0.9.0 release we have been working on ..." and remove "available in trunk".

**Comment by gwenshap** [comment]
Oops, I missed @junrao comments! sorry.
1. Moving Javadoc is a bit more challenging, I'll do it in a separate jira (hopefully I'll have time for it next week)
2. configDef.toHTML no longer has "required", lack of default is what makes something required and we sort accordingly. See KAFKA-2702 for the discussion and patch.

**Comment by gwenshap** [comment]
@guozhangwang: I believe I did mention the new Producer as "Producer" and called the old producer "Old" (see documentation.html)

**Comment by junrao** [comment]
Thanks for the patch. LGTM

**Comment by guozhangwang** [comment]
@gwenshap In the API section, we only list the new producer API as a subsection and left an extra link for the old producer APIs; while in the Configs section, we keep both configs of the new and old producers as separate subsections. So I was wondering if we could be consistent and only list an extra link for old producer configs and remove the subsection.

**Comment by gwenshap** [comment]
We absolutely could modify the API section. Can you open a JIRA for that? It is unrelated to this PR since the API section is not automated. Another option is to completely remove the API section and just link to the Javadoc.

**Pull request closed without merging**